### PR TITLE
Adding "Insert Next" in the SongsList context menu

### PR DESCRIPTION
**Pull request by HinTak**
I found myself wanting this functionality: don't interrupt the current song (unlike "play"), just insert the item after the current playing item. So it differs slightly from "append" (adding to end of playlist) too.

### CODE REVIEW

**Comment by SoongNoonien** [comment]
Thanks for your efforts!
I was thinking about such a feature lately too. But I'm still thinking about some details. I think this option should also be available for whole albums and in the search tab to be consistent. Currently your implementation wouldn't be usable for this since it reverses the order of the file list passed to `files_to_playlist`. Also `filter_to_playlist` needs to be adjusted. But both are minor problems.
I'm also not quite sure if "Insert Next" is the correct name I think "Play Next" is more widely used for this, isn't it? And if we add this feature to albums and the search we need an icon, do you have any suggestions? Perhaps a combination of the play symbol and the plus used for append?

File: src/mpdevil.py [hunk 1]
```diff
@@ -759,7 +761,15 @@ def _to_playlist(self, append, mode):  # modes: play, append, enqueue
 					self.delete(int(duplicates[1]["pos"])-1)
 
 	def files_to_playlist(self, files, mode):
-		def append():
+		def append(next=False):
+			if next:
+				status=self.status()
+				try:
+					for f in files:
+						self.addid(f, int(status["song"])+1)
```
Comment (SoongNoonien) [inline src/mpdevil.py:769]: This is where the implicit reverse happens.

**Comment by SoongNoonien** [comment]
> I think this option should also be available for whole albums and in the search tab to be consistent.

But this is debatable as I consider it to be seldom needed for whole albums.

> I'm also not quite sure if "Insert Next" is the correct name I think "Play Next" is more widely used for this, isn't it?

"As Next" is even short and I think it is still clear enough. I always try to keep the translations in mind and the German translation for "Insert Next" would be something like "Als Nächstes einfügen" which is quite a bit longer than the other entries in the menu.

**Comment by HinTak** [comment]
I know about the reverse order problem, actually. In an earlier cut, I was just adding 'insert next" everywhere next to "append", so I had the whole-album button too. I decided it is probably too much, and also, since the order is reversed, I decided to remove it. Fixing the insert order is not hard.

One thing I found missing, might be too hard to add, is Control/shift-click for multiple selection.

File: src/mpdevil.py (hunk 1 above)
Comment (HinTak) [reply src/mpdevil.py:769]: `for f in files.reverse():` should do it?

Comment (SoongNoonien) [reply src/mpdevil.py:769]: Yes.

**Comment by HinTak** [comment]
The list reversal turns out to be more troublesome - neither `list()`, nor list comprehension (`[s for s in files]`) works. So I have to do it in another list.

Also added (back) the as next button. Don't know how to do the tag filter part, bit started looking at it. (My files aren't tagged so it doesn't help).

**Comment by HinTak** [comment]
Oh, found the problem - `list.reverse()` returns None, and the list itself is changed, it does not return the reversed result. So list comprehension does work, but it isn't possible to do it in one line.

**Comment by SoongNoonien** [comment]
Hi!
Can you see my review comments? I already found the problem with the `list.reverse()` yesterday. It is possible to do it in one line (`for f in reversed(files):`), but we don't need it since mpd can handle this all on its own since 0.23. Take a look at https://mpd.readthedocs.io/en/latest/protocol.html#command-reference. It states regarding `addid` and `findadd` etc.:

> If the second parameter is given, then the song is inserted at the specified position. If the parameter starts with + or -, then it is relative to the current song [8](https://mpd.readthedocs.io/en/latest/protocol.html#since-0-23); e.g. +0 inserts right after the current song and -0 inserts right before the current song (i.e. zero songs between the current song and the newly added song).

**Comment by HinTak** [comment]
Well, that's about the same, because you need to keep a counter: `n = 1 ; insert into (n)th position ; n++`...

**Review by SoongNoonien — CHANGES_REQUESTED** [review]

**Comment by SoongNoonien** [comment]
I looked at the code again and came to the conclusion that I'll probably rewrite the `*_to_playlist` functions. In the process I can easily add the "As Next" function. So don't waste to much time on working on the current state.

File: src/mpdevil.py [hunk 2]
```diff
@@ -736,6 +736,8 @@ def connected(self):
 	def _to_playlist(self, append, mode):  # modes: play, append, enqueue
```
[the remark this answers is not in the document]
Comment (HinTak) [reply src/mpdevil.py:736]: Done

**Comment by HinTak** [comment]
I think the "+0" syntax is on the server-side, and is swallowed by the client. Can't get it work that way. Unless you have a better idea? BTW, I don't have any tag/search in my collection do that part is untested!

**Comment by SoongNoonien** [comment]
Among some other changes I've now reorganized the `*_to_playlist` functions and added the "As Next" feature for single songs. Please take a look at master to see if it behaves as you expected.

> I think the "+0" syntax is on the server-side, and is swallowed by the client.

I don't really understand what you mean by that. Current master uses this feature and at least for me it is working perfectly fine.

> BTW, I don't have any tag/search in my collection so that part is untested!

Just out of curiosity, how are you able to use mpdevil this way? :-)

**Comment by HinTak** [comment]
> Among some other changes I've now reorganized the `*_to_playlist` functions and added the "As Next" feature for single songs. Please take a look at master to see if it behaves as you expected.
> 
> > I think the "+0" syntax is on the server-side, and is swallowed by the client.
> 
> I don't really understand what you mean by that. Current master uses this feature and at least for me it is working perfectly fine.
> 
> > BTW, I don't have any tag/search in my collection so that part is untested!
> 
> Just out of curiosity, how are you able to use mpdevil this way? :-)

Argh, you seems to have drop the whole-album as next button?

Okay, you have the quotes around "+0". That's unclear from the doc - I suppose if you use a number it is converted to a string before sending from the client to the server?

My (intended) use of mpdevil is really as a jukebox/DJ tool - being able the queue songs, and occasionally insert a special one "as next".

**Comment by SoongNoonien** [comment]
> Argh, you seems to have drop the whole-album as next button?

Yes, as you didn't include it in your initial pull request and I'm relatively sure that this function is not actually needed I didn't implement it to keep it simple.

> Okay, you have the quotes around "+0". That's unclear from the doc - I suppose if you use a number it is converted to a string before sending from the client to the server?

Yes, all the arguments are string based. So a simple `+0` without quotes equals to the python integer representing zero which converted to a string is simply `"0"`.

**Comment by HinTak** [comment]
Okay, that's fine. The python-mpd2 doc wasn't clear. I suppose it is obvious that anything on the wire is text, but the doc isn't clear at what stage the conversion happens. Ie. I was afraid it might insist on the input being a number, given that it take "status['song']+1" too.

**Pull request closed without merging**